### ISSUE TITLE
64 bit LuaJIT produces wrong code

### ISSUE THREAD

**Issue by FredyH** [Facepunch/garrysmod-issues#3860]
### Details
It seems that the 64 bit branch of Garry's Mod is using a version of LuaJIT that seems to be extremely broken (at least in certain situations). The two function shown below should evaluate to exactly the same value, but for some reason LuaJIT will "optimize" the first one to never enter the if block, which I think (but I have not verified), has to do something with nans, and LuaJIT doing something dumb with them. 

The obvious explanation would be that LuaJIT sees z == z and changes this condition to always be true. (which would also be incorrect) However, how it decides to instead optimize this condition to always false is beyond me. 

~~Presumably this "optimization" does not happen in the second function because the if statement between the declaration and condition confuses LuaJIT.~~

It appears the result of the function **HEAVILY** depends on the condition that I put into the if, as can be seen in testC(). If you vary the condition it will always produce different results. 

I assume this is because this kind of condition kind of verfifies that z is nan or something and then luajit does some optimizations?

This does **not** happen on 32 bit Garry's Mod.

Using z != z as a condition (and doing sum = sum + 1) has the opposite effect, where the condition is true every single time (for testA and testC).

### Steps to reproduce
Switch to the 64 bit branch of Garry's Mod and run the following program, you will see that the two functions produce different results even though they are functionally equivalent.

```LUA
-- jit.off()  this will fix the bug
-- jit.on()  this will make the bug happen again
local radius = 128
local diameter = radius * 2
	
function testA()
	local sum = 0
	for x2d = 1, 256 do
		for y2d = 1, 256 do
			local x = radius - x2d
			local y = radius - y2d
			local z = math.sqrt(radius * radius - x * x - y * y)
			if (z == z) then -- this makes sure z is not nan
				sum = sum + z
			end
		end
	end
	return sum
end


function testB()
	local sum = 0
	for x2d = 1, 256 do
		for y2d = 1, 256 do
			local x = radius - x2d
			local y = radius - y2d
			local z = math.sqrt(radius * radius - x * x - y * y)
			if (radius * radius - x * x - y * y > 0) then
				
			end
			if (z == z) then
				sum = sum + z
			end
		end
	end
	return sum
end

function testC()
	local sum = 0
	for x2d = 1, 256 do
		for y2d = 1, 256 do
			local x = radius - x2d
			local y = radius - y2d
			local z = math.sqrt(radius * radius - x * x - y * y)
			if (x * x - y * y > 0) then
				
			end
			if (z == z) then
				sum = sum + z
			end
		end
	end
	return sum
end

print(testA(), testB(), testC())
```
Expected output:
> 4392071.1858824      4392071.1858824      4392071.1858824

Output with LuaJIT enabled on 64 bit branch:
> 0      4392071.1858824      18072.250396129

**Comment by Kefta** [comment]
I had a sort-of similar issue in https://github.com/Facepunch/garrysmod-issues/issues/2820, however, it was related to incorrect int sizing (still isn't fixed in JIT 2.1.0). Putting random empty if-statements and local declarations that weren't used around the actual operations did have a similar result of causing the correct result to be compiled under certain conditions, but I was never able to figure out why they would affect the result or why it was inconsistent.

**Comment by willox** [comment]
will be fixed once the update hits steam in a few minutes
